format/udf-spec.md

```diff
@@ -0,0 +1,285 @@
+1. When `secure` is `true`,
+    - Engines **SHOULD NOT** expose the function definition through any inspection (e.g., `SHOW FUNCTIONS`).
+    - Engines **SHOULD** ensure that execution does not leak sensitive information through any channels, such as error messages, logs, or query plans.
+   
+### Overload
+
+Function overloads allow multiple implementations of the same function name with different signatures. Each overload has
+the following fields:
+
+| Requirement | Field name                 | Type                                                   | Description                                                                                   |
+|-------------|----------------------------|--------------------------------------------------------|-----------------------------------------------------------------------------------------------|
+| *required*  | `overload-id`              | `long`                                                 | Monotonically increasing identifier of this function overload.                                |
+| *required*  | `parameters`               | `list<parameter>`                                      | Ordered list of [function parameters](#parameter). Invocation order **must** match this list. |
+| *required*  | `return-type`              | `string` (Iceberg data type)                           | Return type. Example: `"string"`, `"struct<...>"`.                                            |
+| *required*  | `versions`                 | `list<overload-version>`                               | [Versioned implementations](#overload-version) of this overload.                              |
+| *required*  | `current-overload-version` | `long`                                                 | Monotonically increasing identifier of the current overload version.                          |
+| *optional*  | `function-type`            | `string` (either `"udf"` or `"udtf"`, default `"udf"`) | If `"udtf"`, `return-type` must be a `struct` describing the output schema.                   |
+| *optional*  | `doc`                      | `string`                                               | Documentation string.                                                                         |
+
+### Parameter
+| Requirement | Field  | Type                         | Description              |
+|-------------|--------|------------------------------|--------------------------|
+| *required*  | `name` | `string`                     | Parameter name.          |
+| *required*  | `type` | `string` (Iceberg data type) | Parameter data type.     |
+| *optional*  | `doc`  | `string`                     | Parameter documentation. |
+
+Notes:
+1. The `name` and `type` of a `parameter` are immutable. To change them, a new overload must be created. Only the optional documentation field (`doc`) can be updated in-place.
+2. The `return-type` is immutable. To change it, users must create a new overload and deprecate or remove the old one.
```

Comment:
What if the return type generated by one engine is different? For instance, what if I have an `index_of` function that looks for a value in a list. In one engine, lists are limited to < Integer.MAX_VALUE and in another they are not so the engines produce `int` and `long` return values. I think it would be okay to promote a definition that produces `int` to `long`. I'm also not sure how you would do it without replacing the value. Otherwise you'd have multiple definitions with the same parameters but different return types.

I think we want to avoid allowing different overloads/definitions with the same parameters right?